File: Models/Statistics.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Bakery.Models
{
    public class Statistics<T>
    {
        public T Object { get; set; }
        public int NumberOfOrders { get; set; }
        public string ObjectName { get; set; }
        public List<Statistics<Cake>> OrderedTogetherCakes { get; set; } = new List<Statistics<Cake>>();
        public List<Statistics<Cream>> OrderedTogetherCreams { get; set; } = new List<Statistics<Cream>>();
        public List<Statistics<Additional>> OrderedTogetherAdditionals { get; set; } = new List<Statistics<Additional>>();
        public List<Statistics<Decoration>> OrderedTogetherDecorations { get; set; } = new List<Statistics<Decoration>>();
    }
}

File: ClientApp/src/app/models/Statistics.ts
import { Additional } from "./Additional";
import { Cake } from "./Cake";
import { Cream } from "./Cream";
import { Decoration } from "./Decoration";

export class Statistics<T> {
    object: T;
    numberOfOrders: number;
    objectName: string;
    orderedTogetherCakes: Statistics<Cake>[] = [];
    orderedTogetherCreams: Statistics<Cream>[] = [];
    orderedTogetherAdditionals: Statistics<Additional>[] = [];
    orderedTogetherDecorations: Statistics<Decoration>[] = [];
}